### ISSUE TITLE
Refactor game state specific things into new GameState structure.

### ISSUE THREAD

**Issue by ZehMatt** [OpenRCT2/OpenRCT2#21193]

**Comment by Gymnasiast** [comment]
I’d like to “reserve” these:
- gGuestInitialCash
- gGuestInitialHappiness
- gGuestInitialHunger
- gGuestInitialThirst

**Comment by Harry-Hopkinson** [comment]
I could do all of the climate ones to help out.

- gClimate
- gClimateCurrent
- gClimateNext
- gClimateUpdateTimer

**Comment by Gymnasiast** [comment]
I’d like to reserve:
- gWeeklyProfitAverageDividend
- gWeeklyProfitAverageDivisor
- gWeeklyProfitHistory

**Comment by Broxzier** [comment]
I'd like to reserve the remaining `gPark...` globals:

- gParkEntranceFee
- gParkEntrances
- gParkRatingCasualtyPenalty
- gParkSize
- gParkValue
- gParkValueHistory

**Comment by Gymnasiast** [comment]
I’d like to reserve the gScenario* globals.

**Comment by Harry-Hopkinson** [comment]
I would like to reserve the following guest related globals...

- gNextGuestNumber
- gNumGuestsHeadingForPark
- gNumGuestsInPark
- gNumGuestsInParkLastWeek

**Comment by Harry-Hopkinson** [comment]
I could do the following income/admissions states...

- gTotalAdmissions
- gTotalIncomeFromAdmissions
 - gTotalRideValueForMoney

**Comment by Harry-Hopkinson** [comment]
I could do the following:

 - gMapBaseZ
 - gMapSize
 - gMaxBankLoan
 - gNewsItems

**Comment by Harry-Hopkinson** [comment]
I could do the following:

- gSavedAge
- gSavedView
- gSavedViewRotation
- gSavedViewZoom

**Comment by Martymoose98** [comment]
I could refractor:

- gBankLoan
- gBankLoanInterestRate

**Comment by Harry-Hopkinson** [comment]
I could do the following:

- gCompanyValue
- gConstructionRightsPrice
- gCurrentExpenditure
- gCurrentProfit

**Comment by Harry-Hopkinson** [comment]
I can do:

- gEditorStep
- gExpenditureTable
- gGuestChangeModifier

**Comment by jan-kelemen** [comment]
I did
- gCashHistory

I would like to reserve
- gHistoricalProfit
- gGuestsInParkHistory

**Comment by jan-kelemen** [comment]
I'll do
- gRideRatingUpdateStates
- gSamePriceThroughoutPark

**Comment by Harry-Hopkinson** [comment]
I can do:

- gLandPrice
- gLastEntranceStyle

**Comment by Gymnasiast** [comment]
@ZehMatt Shouldn’t the following variables be in here as well?

gEntityLists
_freeIdList
_entityFlashingList
gEntitySpatialIndex

**Comment by ZehMatt** [comment]
> @ZehMatt Shouldn’t the following variables be in here as well?
> 
> gEntityLists _freeIdList _entityFlashingList gEntitySpatialIndex

No, they are not saved in the file and are typically rebuilt upon loading it.

**Comment by jan-kelemen** [comment]
Done the `gWidePathTileLoopPosition`, I've also noticed that `gGrassSceneryTileLoopPosition` along side it that is saved in the file, so done that too.

**Comment by Gymnasiast** [comment]
> > @ZehMatt Shouldn’t the following variables be in here as well?
> > gEntityLists _freeIdList _entityFlashingList gEntitySpatialIndex
> 
> No, they are not saved in the file and are typically rebuilt upon loading it.

I know they’re not saved in the file, but they are still gamestate-specific. That is to say, they are specific to the currently loaded game and using the list/index of one park on the other will cause severe problems.

**Comment by ZehMatt** [comment]
> > > @ZehMatt Shouldn’t the following variables be in here as well?
> > > gEntityLists _freeIdList _entityFlashingList gEntitySpatialIndex
> > 
> > 
> > No, they are not saved in the file and are typically rebuilt upon loading it.
> 
> I know they’re not saved in the file, but they are still gamestate-specific. That is to say, they are specific to the currently loaded game and using the list/index of one park on the other will cause severe problems.

Yeah, we need to refactor that part a bit better but they don't really belong in the game state, that is more a system specific state. I would like to split those things apart so that the entity spatial index is its own thing and not part of the entity registry, the tiles mapping should be also separated. The best way would be to have system based events, example: Entity registry notifies the Spatial mapping that an entity is removed, moving an entity also notifies the spatial system and so on, we need to better separate responsibilities in general.

**Comment by Harry-Hopkinson** [comment]
I can do the following:

- gPeepSpawns
- gPeepWarningThrottle

**Comment by Harry-Hopkinson** [comment]
I can do:

- gWidePathTileLoopPosition

**Comment by jan-kelemen** [comment]
> I can do:
> 
> * gWidePathTileLoopPosition

@Harry-Hopkinson there is already a PR for that https://github.com/OpenRCT2/OpenRCT2/pull/21554

**Comment by Harry-Hopkinson** [comment]
Sorry about that, did not realise.

**Comment by Gymnasiast** [comment]
@Harry-Hopkinson You can also do `gParkRatingCasualtyPenalty`, AFAIC. It is assigned by @Broxzier , but he is probably busy as he never made a PR for it.

**Comment by tupaschoal** [comment]
Is that it, have we finished moving them All?

**Comment by Gymnasiast** [comment]
Hm, _maybe_. There is a similarly needed structure called `GameState` which contains some park-related stuff and the date, which I would assume needs to be absorbed into GameState_t. But I’d like @ZehMatt to confirm that.

**Comment by ZehMatt** [comment]
> Hm, _maybe_. There is a similarly needed structure called `GameState` which contains some park-related stuff and the date, which I would assume needs to be absorbed into GameState_t. But I’d like @ZehMatt to confirm that.

Yea, all of the data it currently holds should be moved into GameState_t, minor oversight due to how those classes hide the data.

**Comment by Gymnasiast** [comment]
Yes, that seems appropriate. I updated the initial post.

**Comment by AaronVanGeffen** [comment]
@ZehMatt Should the tile elements become part of the game state struct?

**Comment by ZehMatt** [comment]
> @ZehMatt Should the tile elements become part of the game state struct?

Yea

**Comment by AaronVanGeffen** [comment]
Ah, it seems @Gymnasiast already took care of that in #21454. I think we can start working towards asynchronous loading, then, right.

**Comment by AaronVanGeffen** [comment]
I believe the following also need moving into `GameState_t`:

Scenario.h/cpp:
```
std::string gScenarioSavePath;
bool gFirstTimeSaving;
uint32_t gLastAutoSaveUpdate;

std::string gScenarioFileName;
```

EntityRegistry.cpp:
```cpp
static std::array<std::list<EntityId>, EnumValue(EntityType::Count)> gEntityLists;
static std::vector<EntityId> _freeIdList;

static bool _entityFlashingList[MAX_ENTITIES];

static std::array<std::vector<EntityId>, SPATIAL_INDEX_SIZE> gEntitySpatialIndex;
``

**Comment by duncanspumpkin** [comment]
The entity ones are recreated on load they work a bit differently to normal game state things

**Comment by AaronVanGeffen** [comment]
I believe that, but they shouldn't be overwritten if we happen to have several game states around, right.

**Comment by Harry-Hopkinson** [comment]
Do the following variables need moving into the GameState structure?

```
std::string gScenarioSavePath;
bool gFirstTimeSaving;
uint32_t gLastAutoSaveUpdate;

std::string gScenarioFileName;
```

if so, I could move them over to help out.

**Comment by duncanspumpkin** [comment]
> Do the following variables need moving into the GameState structure?
> 
> ```
> std::string gScenarioSavePath;
> bool gFirstTimeSaving;
> uint32_t gLastAutoSaveUpdate;
> 
> std::string gScenarioFileName;
> ```
> 
> if so, I could move them over to help out.

Yup go for it

**Comment by ZehMatt** [comment]
> > Do the following variables need moving into the GameState structure?
> > ```
> > std::string gScenarioSavePath;
> > bool gFirstTimeSaving;
> > uint32_t gLastAutoSaveUpdate;
> > 
> > std::string gScenarioFileName;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > if so, I could move them over to help out.
> 
> Yup go for it

I think the first 3 variables don't belong in there, those should not be saved in the save file other than the name.

**Comment by Gymnasiast** [comment]
Sure, but they’re still part of the game state, even if they’re not saved in the file, aren’t they?

**Comment by ZehMatt** [comment]
> Sure, but they’re still part of the game state, even if they’re not saved in the file, aren’t they?

The only role those variables play are when and to where the game state is saved but they are not really part of it, that is data for what I would call the autosave system.

**Comment by Harry-Hopkinson** [comment]
So should I only move gScenarioFileName into the GameState structure for this PR https://github.com/OpenRCT2/OpenRCT2/pull/22662 then?

**Comment by ZehMatt** [comment]
> So should I only move gScenarioFileName into the GameState structure for this PR #22662 then?

That would be good. 

To clarify what goes into the game state struct, it can be answered with asking the question: Can the game advance one tick without the variable? If the answer is yes then it doesn't belong in there.

**Comment by Harry-Hopkinson** [comment]
I have rebased and remove those commits from the PR.
